pkg/promotion/releasepromotion/release.go

```diff
@@ -0,0 +1,205 @@
+			return os.Remove(path)
+		}
+		return nil
+	}); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	// add names back
+	for key, target := range v.targets {
+		base := filepath.Join(append([]string{toDir}, key.Directory()...)...)
+		if err := os.MkdirAll(base, 0755); err != nil {
+			return err
+		}
+		keys := sets.NewString()
+		for name := range target {
+			keys.Insert(name)
+		}
+		buf := &bytes.Buffer{}
+		fmt.Fprintf(buf, "# generated by %s, do not edit\n", commandName)
+
+		for _, name := range keys.List() {
+			src := target[name].info
+			fmt.Fprintf(buf, "%s from %s promotes to %s\n", name, src.RelativePath(), key.String())
+		}
+		if err := ioutil.WriteFile(filepath.Join(base, "images.txt"), buf.Bytes(), 0640); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type verifierMap map[promotionTarget]promotionNames
+
+func (m verifierMap) get(key promotionTarget) promotionNames {
+	v, ok := m[key]
+	if ok {
+		return v
+	}
+	v = make(promotionNames)
+	m[key] = v
+	return v
+}
+
+type promotionTarget struct {
+	Namespace  string
+	Name       string
+	NamePrefix string
+	Tag        string
+}
+
+func (t promotionTarget) Empty() bool {
+	return len(t.Namespace) == 0 || (len(t.Name) == 0 && len(t.NamePrefix) == 0 && len(t.Tag) == 0)
+}
+
+func (t promotionTarget) String() string {
+	if len(t.NamePrefix) > 0 {
+		return fmt.Sprintf("%s/%s{image}:%s", t.Namespace, t.NamePrefix, t.Tag)
+	}
+	if len(t.Name) > 0 {
+		return fmt.Sprintf("%s/%s:{image}", t.Namespace, t.Name)
+	}
+	return fmt.Sprintf("%s/{image}:%s", t.Namespace, t.Tag)
+}
+
+func (t promotionTarget) Directory() []string {
+	if len(t.NamePrefix) > 0 {
+		return []string{t.Namespace, t.NamePrefix, t.Tag}
+	}
+	if len(t.Name) > 0 {
+		return []string{t.Namespace, t.Name}
+	}
+	return []string{t.Namespace, "tag", t.Tag}
+}
+
+func (t promotionTarget) PullSpec() string {
+	return fmt.Sprintf("{registry}/%s", t.String())
+}
+
+type promotionNames map[string]promotionSource
+
+func (m promotionNames) addReference(from, to string, included sets.String, info *config.Info, target promotionTarget) error {
+	if !included.Has(from) && from != "src" && from != "rpms" {
+		return fmt.Errorf("promotes an image %q that is not built with via a project directory image build step or rpm step", from)
+	}
+	if included.Has(to) && from == to {
+		return nil
+	}
+	source := promotionSource{info: info}
+	if existing, ok := m[to]; ok {
+		return fmt.Errorf("%s is already promoting an image named %q to %s", existing.info.Filename, from, target)
+	}
+	m[to] = source
+	return nil
+}
+
+func (m promotionNames) addStep(image *cioperatorapi.ProjectDirectoryImageBuildStepConfiguration, optional bool, excluded, included sets.String, info *config.Info, target promotionTarget) error {
+	if image == nil {
+		return nil
+	}
+	name := string(image.To)
+	if len(name) == 0 {
+		return nil
+	}
+	if _, ok := included[string(name)]; ok {
+		return fmt.Errorf("defines image %q more than once", name)
```

Comment:
Similarly, defensive programming is fine but you could jettison all of this since they are invariants that are enforced elsewhere